contracts/core/EntryPoint.sol

```diff
@@ -698,7 +715,15 @@ contract EntryPoint is IEntryPoint, StakeManager, NonceManager, ReentrancyGuard,
 
             actualGasCost = actualGas * gasPrice;
             if (opInfo.prefund < actualGasCost) {
-                revert FailedOp(opIndex, "AA51 prefund below actualGasCost");
+                if (mode == IPaymaster.PostOpMode.postOpReverted) {
+                    emitPrefundTooLow(actualGas, opInfo);
+                    return opInfo.prefund;
+                } else {
+                    assembly ("memory-safe") {
+                        mstore(0, INNER_REVERT_LOW_PREFUND)
+                        revert(0, 32)
+                    }
+                }
             }
             uint256 refund = opInfo.prefund - actualGasCost;
             _incrementDeposit(refundAddress, refund);
```

Comment:
this code about paying gas and emitting event can be moved to '_executeUserOp'

Reply:
the entire "_postExecution" is about event and refund